packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -1868,6 +1871,212 @@ function commitPassiveLifeCycles(finishedWork: Fiber): void {
+        // This is a mount
+        if (current === null) {
+          if (effectTag & Update) {
+            switch (fiber.tag) {
+              case FunctionComponent:
+              case ForwardRef:
+              case SimpleMemoComponent:
+              case Block: {
+                invokeGuardedCallback(
+                  null,
+                  commitHookEffectListUnmount,
+                  null,
+                  HookLayout | HookHasEffect,
+                  fiber,
+                );
+                if (hasCaughtError()) {
+                  const unmountError = clearCaughtError();
+                  captureCommitPhaseError(fiber, unmountError);
+                }
+                break;
+              }
+              case ClassComponent: {
+                const instance = fiber.stateNode;
+                if (typeof instance.componentWillUnmount === 'function') {
+                  safelyCallComponentWillUnmount(fiber, instance);
+                }
+                break;
+              }
+            }
+          }
+        }
+        fiber = fiber.sibling;
+      }
+    }
+  }
+}
+
+function invokeLayoutEffectsMountInDEV(firstChild) {
+  if (enableDoubleInvokingEffects) {
+    // mount layout effects
+    if (__DEV__) {
+      let fiber = firstChild;
+      while (fiber !== null) {
+        if (fiber.child !== null) {
+          // Should we add a separate subtree tag for this?
+          const primarySubtreeTag = fiber.subtreeTag & LayoutSubtreeTag;
+          if (primarySubtreeTag !== NoSubtreeTag) {
+            invokeLayoutEffectsMountInDEV(fiber.child);
+          }
+        }
+
+        const effectTag = fiber.effectTag;
+        const current = fiber.alternate;
+        if (current === null) {
+          if (effectTag & Update) {
+            switch (fiber.tag) {
+              case FunctionComponent:
+              case ForwardRef:
+              case SimpleMemoComponent:
+              case Block: {
+                invokeGuardedCallback(
+                  null,
+                  commitHookEffectListMount,
+                  null,
+                  HookLayout | HookHasEffect,
+                  fiber,
+                );
+                if (hasCaughtError()) {
+                  const mountError = clearCaughtError();
+                  captureCommitPhaseError(fiber, mountError);
+                }
+                break;
+              }
+              case ClassComponent: {
+                const instance = fiber.stateNode;
+                instance.componentDidMount();
+                break;
+              }
+            }
+          }
+        }
+        fiber = fiber.sibling;
+      }
+    }
+  }
+}
+
+function invokePassiveEffectsUnmountInDEV(firstChild): void {
+  if (enableDoubleInvokingEffects) {
+    if (__DEV__) {
+      let fiber = firstChild;
+      while (fiber !== null) {
+        if (fiber.child !== null) {
+          const primarySubtreeTag = fiber.subtreeTag & PassiveSubtreeTag;
```

Comment:
Seems like the same over-traversal stuff is possible here too (so we maybe would want the same TODO comment?)